app/src/data/host-context.tsx

```diff
@@ -0,0 +1,125 @@
+        loading: boolean
+        message: string
+    }
+}
+
+enum HostDashboardActionType {
+    BeginFetchQuestions,
+    FinishFetchQuestions,
+    GetHostById,
+    isLoading,
+    BeginPostAnswer,
+    FinishPostAnswer,
+    Error,
+}
+
+interface HostDashboardAction {
+    type: HostDashboardActionType
+    payload?: HostDashboardData | Array<QuestionType> | boolean | string
+}
+
+const initialState: HostDashboardData = {
+    hostQuestions: [],
+    hostResponses: [],
+    loaderState: {
+        loading: false,
+        message: '',
+    },
+}
+
+function hostDashboardReducer(
+    state: HostDashboardData,
+    action: HostDashboardAction
+): HostDashboardData {
+    switch (action.type) {
+        case HostDashboardActionType.BeginFetchQuestions: {
+            return {
+                ...state,
+                loaderState: {
+                    loading: true,
+                    message: action.payload as string,
+                },
+            }
+        }
+        case HostDashboardActionType.FinishFetchQuestions: {
+            return {
+                ...state,
+                loaderState: {
+                    loading: false,
+                    message: 'Finished loading',
+                },
+                hostQuestions: action.payload as Array<QuestionType>,
+            }
+        }
+        default:
+            throw new Error(`Unsupported action: ${JSON.stringify(action)}`)
+    }
+}
+
+export function HostDashboardDataProvider(
+    props: React.PropsWithChildren<{}>
+): JSX.Element {
+    const [state, dispatch] = React.useReducer(
+        hostDashboardReducer,
+        initialState
+    )
+
+    //dispatch begin and finish get questions
+
+    React.useEffect(() => {
+        ;(async function () {
+            console.log('loadData: fetching...')
+            try {
+                dispatch({
+                    type: HostDashboardActionType.BeginFetchQuestions,
+                    payload: 'loading',
+                })
+                const questions = await hostsFetcher.getHostQuestions()
+
+                dispatch({
+                    type: HostDashboardActionType.FinishFetchQuestions,
+                    payload: questions,
+                })
+            } catch (e) {
+                dispatch({
+                    type: HostDashboardActionType.Error,
+                    payload: `System error: ${e}`,
+                })
+            }
+        })()
+    }, [])
+
+    const value = React.useMemo(() => [state, dispatch], [state])
+    return <HostDashboardContext.Provider value={value} {...props} />
+}
+
+export function useHostDashboardData() {
+    const context = React.useContext(HostDashboardContext)
+    if (!context) {
+        throw new Error(
+            'useHostDashboardData must be used within a HostDashboardProvider'
+        )
+    }
+
+    const [data, dispatch] = context as [
+        HostDashboardData,
+        React.Dispatch<HostDashboardAction>
+    ]
+
+    return {
```

Comment:
Let's discuss what abstractions will be the most useful for driving state from lower-level components. Ideally, a consumer does not need to understand the action and payload types to use this hook via the `dispatch` function and we can provide more descriptive callbacks which encapsulate the use of dispatch.

Reply:
yes absolutely. i did not expand any further because I still feel unclear on business logic.